File: modules/graphtraversal/TopologicalSort.java
/*
 * Topological Sort  (CSC310 Module 7 - Graph Algorithms)
 * ======================================================
 *
 * DFS method: output vertices in DECREASING finish time (prepend on finish).
 * A back edge (a GRAY neighbor) means the graph has a cycle and no order exists.
 *
 *   javac TopologicalSort.java && java TopologicalSort
 *
 * Time: O(|V| + |E|).
 */
import java.util.*;

public class TopologicalSort {

    static final int WHITE = 0, GRAY = 1, BLACK = 2;
    static Map<String, Integer> color = new HashMap<>();
    static Deque<String> order = new ArrayDeque<>();   // prepend on finish
    static boolean hasCycle = false;

    static void dfsVisit(String u, Map<String, List<String>> adj) {
        color.put(u, GRAY);
        for (String v : adj.get(u)) {
            int c = color.get(v);
            if (c == WHITE) dfsVisit(v, adj);
            else if (c == GRAY) hasCycle = true;       // back edge -> cycle
        }
        color.put(u, BLACK);
        order.addFirst(u);
    }

    static List<String> topologicalSort(Map<String, List<String>> adj) {
        for (String u : adj.keySet()) color.put(u, WHITE);
        for (String u : adj.keySet())
            if (color.get(u) == WHITE) dfsVisit(u, adj);
        return hasCycle ? null : new ArrayList<>(order);
    }

    public static void main(String[] args) {
        // CLRS "getting dressed" DAG (not used in any CSC310 assignment).
        Map<String, List<String>> adj = new LinkedHashMap<>();
        adj.put("undershorts", List.of("pants", "shoes"));
        adj.put("pants",       List.of("belt", "shoes"));
        adj.put("belt",        List.of("jacket"));
        adj.put("shirt",       List.of("belt", "tie"));
        adj.put("tie",         List.of("jacket"));
        adj.put("jacket",      List.of());
        adj.put("socks",       List.of("shoes"));
        adj.put("shoes",       List.of());
        adj.put("watch",       List.of());

        List<String> result = topologicalSort(adj);
        if (result == null) {
            System.out.println("Graph has a cycle: no topological ordering exists.");
        } else {
            System.out.println("A valid topological order:");
            System.out.println("  " + String.join(" -> ", result));
        }
    }
}

File: modules/graphtraversal/topological_sort.py
"""
Topological Sort  (CSC310 Module 7 - Graph Algorithms)
======================================================

A topological sort of a directed acyclic graph (DAG) is a linear ordering of the
vertices such that for every edge (u, v), u appears before v.  The DFS method
from the notes: run DFS and output vertices in DECREASING order of finish time
(prepend each vertex as it finishes).  A back edge (a GRAY neighbor) means the
graph has a cycle and no topological order exists.

Time: O(|V| + |E|).
"""

WHITE, GRAY, BLACK = 0, 1, 2


def topological_sort(adj):
    color = {u: WHITE for u in adj}
    order = []           # vertices prepended in reverse finish order
    has_cycle = [False]

    def dfs_visit(u):
        color[u] = GRAY
        for v in adj[u]:
            if color[v] == WHITE:
                dfs_visit(v)
            elif color[v] == GRAY:      # back edge -> cycle
                has_cycle[0] = True
        color[u] = BLACK
        order.insert(0, u)              # finished: prepend

    for u in adj:
        if color[u] == WHITE:
            dfs_visit(u)
    return (None if has_cycle[0] else order)


def main():
    # CLRS "getting dressed" DAG -- a classic dependency example (not used in any
    # CSC310 assignment).  Edge u -> v means "u must be put on before v".
    adj = {
        "undershorts": ["pants", "shoes"],
        "pants":       ["belt", "shoes"],
        "belt":        ["jacket"],
        "shirt":       ["belt", "tie"],
        "tie":         ["jacket"],
        "jacket":      [],
        "socks":       ["shoes"],
        "shoes":       [],
        "watch":       [],
    }
    order = topological_sort(adj)
    print("Dependency edges (u must come before v):")
    for u in adj:
        for v in adj[u]:
            print(f"  {u} -> {v}")
    print()
    if order is None:
        print("Graph has a cycle: no topological ordering exists.")
    else:
        print("A valid topological order:")
        print("  " + " -> ".join(order))


if __name__ == "__main__":
    main()
